### DISCUSSION TITLE
Refresh UI Debugger from plugin

### DISCUSSION THREAD

**Discussion by mfthomps** [NationalSecurityAgency/ghidra#6295] [Debugger]
Please refer to closed issue #4174
I'm back to this same question for Ghidra11, which no longer has getDebuggerObjectsProvider. I have tried:

                    dbo = (DebuggerObjectsProvider) tool.getComponentProvider("Objects");
                    dbo.refresh("Threads");
But that does not do the refresh. I notice that in the Objects window of the debugger, selecting "Threads" and doing "refresh" does not work either. I must select "Stack" of the thread. Any suggestions on how to do this with in my plugin?

**Comment by nsadeveloper789** [comment]
I'm not sure what has changed in 11 that caused this issue. Still, I think the best way to do this is to access the model API directly, as in this response: https://github.com/NationalSecurityAgency/ghidra/issues/4174#issuecomment-1144926259. The UI installs listeners on that model, so if anything asks the model to refresh, the UI will update, too. The posted example tries to capture a lot of alternatives. Let me know if you'd like some help paring it down.

**Comment by mfthomps** [comment]
I previously implemented that suggestion within my plugin in order to get the registers to refresh.  And that worked fine, and that still works fine.  However the traces do not refresh.   For example, when I skip to a new execution point with my plugin, the registers refresh but the traces do not refresh until I select the Stack object and refresh.

**Comment by mfthomps** [comment]
Ah, got it.  Needed to change the dynamic trace to "Track Program Counter (by Register)"
Any way to set that as a default in my plugin?

**Reply by nsadeveloper789** [reply]
A couple alternatives here:
1. The "Track Program Counter" option will try the stack first, then registers. It only falls back to registers if the stack is completely absent. In your case, it's just stale. You can use the same refresh technique as for registers, but with `TargetStack` instead of `TargetRegisterBank`, to ensure the stacks are all fresh, too. (This generalizes to any class/interface extending from `TargetObject`.)
2. If you'd rather just always use registers for tracking, then you can set the listing's tracking spec:
`tool.getService(DebuggerListingService.class).setTrackingSpec(PCByRegisterLocationTrackingSpec.INSTANCE)`. That'll override the user's setting (which is saved from session to session) whenever you call it, though. It doesn't just change the "default."

**Comment by mfthomps** [comment]
Thanks for those solutions.  I appreciate it!
